/**
* Hello World API
* Say hello API
*
* The version of the OpenAPI document: 1.0.0
* 
*
* NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
* https://openapi-generator.tech
* Do not edit the class manually.
*/
/*
 * HelloServiceApi.h
 *
 * 
 */

#ifndef HelloServiceApi_H_
#define HelloServiceApi_H_


#include <pistache/http.h>
#include <pistache/router.h>
#include <pistache/http_headers.h>

#include <optional>
#include <utility>

#include "Error.h"
#include "Hello.h"
#include <string>

namespace org::openapitools::server::api
{

class  HelloServiceApi {
public:
    explicit HelloServiceApi(const std::shared_ptr<Pistache::Rest::Router>& rtr);
    virtual ~HelloServiceApi() = default;
    void init();

    static const std::string base;

private:
    void setupRoutes();

    void get_hello_handler(const Pistache::Rest::Request &request, Pistache::Http::ResponseWriter response);
    void get_hello_by_lang_handler(const Pistache::Rest::Request &request, Pistache::Http::ResponseWriter response);
    void update_hello_handler(const Pistache::Rest::Request &request, Pistache::Http::ResponseWriter response);
    void hello_service_api_default_handler(const Pistache::Rest::Request &request, Pistache::Http::ResponseWriter response);

    const std::shared_ptr<Pistache::Rest::Router> router;

    /// <summary>
    /// Helper function to handle unexpected Exceptions during Parameter parsing and validation.
    /// May be overridden to return custom error formats. This is called inside a catch block.
    /// Important: When overriding, do not call `throw ex;`, but instead use `throw;`.
    /// </summary>
    virtual std::pair<Pistache::Http::Code, std::string> handleParsingException(const std::exception& ex) const noexcept;

    /// <summary>
    /// Helper function to handle unexpected Exceptions during processing of the request in handler functions.
    /// May be overridden to return custom error formats. This is called inside a catch block.
    /// Important: When overriding, do not call `throw ex;`, but instead use `throw;`.
    /// </summary>
    virtual std::pair<Pistache::Http::Code, std::string> handleOperationException(const std::exception& ex) const noexcept;

    /// <summary>
    /// Get hello
    /// </summary>
    /// <remarks>
    /// 
    /// </remarks>
    virtual void get_hello(Pistache::Http::ResponseWriter &response) = 0;
    /// <summary>
    /// Get hello by language
    /// </summary>
    /// <remarks>
    /// 
    /// </remarks>
    /// <param name="language"></param>
    virtual void get_hello_by_lang(const std::string &language, Pistache::Http::ResponseWriter &response) = 0;
    /// <summary>
    /// update the way to say hello.
    /// </summary>
    /// <remarks>
    /// 
    /// </remarks>
    virtual void update_hello(Pistache::Http::ResponseWriter &response) = 0;

};

} // namespace org::openapitools::server::api

#endif /* HelloServiceApi_H_ */

